docs/guide/introduction/configuration.md
# Configuration

Though this package is crafted to suit most of your needs by default, you can edit the configuration file to suit
certain demands.

## Environment

| Name                                   | Description                          | Default            | 
|----------------------------------------|--------------------------------------|--------------------|
| `WALLET_CONSISTENCY_CHECK_ENABLED`     | Weather to enable consistency check. | true               |
| `WALLET_CONSISTENCY_SECRET`            | Secret key for checksum generation.  | consistency_secret |
| `WALLET_WALLET_TABLE_NAME`             | Wallet table name                    | wallets            |
| `WALLET_DEFAULT_WALLET_NAME`           | Default wallet name                  | Default Wallet     |
| `WALLET_DEFAULT_WALLET_SLUG`           | Default wallet slug                  | default            |
| `WALLET_DEFAULT_WALLET_DECIMAL_PLACES` | Default wallet decimal places        | 24                 |
| `WALLET_TRANSACTION_TABLE_NAME`        | Transaction table name               | transactions       |
| `WALLET_CACHE_DRIVER`                  | Cache for wallet balance             | array              |
| `WALLET_CACHE_TTL`                     | Cache TTL for wallet balance         | 24h                |
| `WALLET_LOCK_DRIVER`                   | Lock for wallets                     | array              |
| `WALLET_LOCK_TTL`                      | Lock TTL for wallets                 | 1s                 |
| `WALLET_MATH_SCALE`                    | Select mathematical precision        | 64                 |

## Configure default wallet

Customize `name`,`slug`, `meta` and `decimalPlaces` of default wallet.

config/wallet.php:

```php
'default' => [
    'name' => 'Ethereum',
    'slug' => 'ETH',
    'meta' => [],
    'decimal_places' => 18,
],
```

## Extend base Wallet model

You can extend base Wallet model by creating a new class that extends `AliRaghebi\Wallet\Models\Wallet` and registering the
new class in `config/wallet.php`.
Example `MyWallet.php`

App/Models/MyWallet.php:

```php
use AliRaghebi\Wallet\Models\Wallet as WalletBase;

class MyWallet extends WalletBase {
    public function helloWorld(): string { return "hello world"; }
}
```

### Register base Wallet model

config/wallet.php:

```php
'wallet' => [
    'table' => 'wallets',
    'model' => MyWallet::class,
    'default' => [
        'name' => 'Default Wallet',
        'slug' => 'default',
        'meta' => [],
    ],
],
```

```php
echo $user->wallet->helloWorld();
```

This same method above, can be used to extend the base `Transaction` model and registering the extended
model in the configuration file.



config/config.php
<?php

use AliRaghebi\Wallet\Models\Transaction;
use AliRaghebi\Wallet\Models\Transfer;
use AliRaghebi\Wallet\Models\Wallet;

return [
    /*
    |--------------------------------------------------------------------------
    |  Number digits and decimal places used in migrations and calculations
    |--------------------------------------------------------------------------
    */
    'number' => [
        /*
        | Total number of digits used to store numbers in the database, including both integer and fractional parts.
        */
        'digits' => 64,

        /*
        | Number of decimal digits used for storing and calculating values.
        | All numeric values will be scaled to the specified `decimal_places` to ensure data integrity and consistency.
        */
        'decimal_places' => env('WALLET_DECIMAL_PLACES', 24),
    ],

    /*
    |--------------------------------------------------------------------------
    |  Wallet integrity check settings
    |--------------------------------------------------------------------------
    |
    | Integrity is verified using checksum, in case of checksum mismatch, transaction will be failed and rolled back.
    */
    'integrity_validation' => [
        /*
        | Weather to enable integrity check. If you disable this option, no validation will be done on wallets on update.
        */
        'enabled' => env('WALLET_INTEGRITY_VALIDATION_ENABLED', true),
        /*
        | Secret key for checksum generation.
        */
        'secret' => env('WALLET_INTEGRITY_VALIDATION_SECRET', 'integrity_validation_secret'),
    ],

    /*
    |--------------------------------------------------------------------------
    |  Base model 'wallet'.
    |--------------------------------------------------------------------------
    |
    | Contains the configuration for the wallet model.
    */
    'wallet' => [
        /*
        | The table name for wallets.
        |
        | This value is used to store wallets in a database.
        */
        'table' => env('WALLET_WALLET_TABLE_NAME', 'wallets'),

        /*
        | The model class for wallets.
        |
        | This value is used to create new wallets.
        */
        'model' => Wallet::class,

        /*
        | The default configuration for wallets.
        */
        'default' => [
            /*
            | The name of the default wallet.
            */
            'name' => env('WALLET_DEFAULT_WALLET_NAME', 'Default Wallet'),

            /*
            | The slug of the default wallet.
            */
            'slug' => env('WALLET_DEFAULT_WALLET_SLUG', 'default'),

            /*
            | The meta information of the default wallet.
            */
            'meta' => null,
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    |  Base model 'transaction'.
    |--------------------------------------------------------------------------
    */
    'transaction' => [
        /*
        | The table name for transactions.
        |
        | This value is used to store transactions in a database.
        */
        'table' => env('WALLET_TRANSACTION_TABLE_NAME', 'transactions'),

        /*
        | The model class for transactions.
        |
        | This value is used to create new transactions.
        */
        'model' => Transaction::class,
    ],

    /*
   |--------------------------------------------------------------------------
   |  Base model 'transfer'.
   |--------------------------------------------------------------------------
   */
    'transfer' => [
        /*
        | The table name for transfers.
        |
        | This value is used to store transfers in a database.
        */
        'table' => env('WALLET_TRANSFER_TABLE_NAME', 'transfers'),

        /*
        | The model class for transfer.
        |
        | This value is used to create new transfer.
        */
        'model' => Transfer::class,
    ],

    /*
    |--------------------------------------------------------------------------
    |  Storage of the state of the balance of wallets.
    |--------------------------------------------------------------------------
    |
    | This is used to cache the results of calculations
    | in order to improve the performance of the package.
    |
    */
    'cache' => [
        /*
        | The driver for the cache.
        */
        'driver' => env('WALLET_CACHE_DRIVER', 'array'),
        /*
        | The time to live for the cache in seconds.
        */
        'ttl' => env('WALLET_CACHE_TTL', 24 * 3600),
    ],

    /*
    |--------------------------------------------------------------------------
    |  A system for dealing with race conditions.
    |--------------------------------------------------------------------------
    |
    | This is used to protect against race conditions
    | when updating the balance of a wallet.
    |
    */
    'lock' => [
        /*
        | The driver for the lock.
        |
        | The following drivers are supported:
        | - array
        | - redis
        | - memcached
        | - database
        */
        'driver' => env('WALLET_LOCK_DRIVER', 'array'),

        /*
        | The time to live for the lock in seconds.
        */
        'seconds' => env('WALLET_LOCK_TTL', 1),
    ],
];
